### PR TITLE
First attempt at adding --port-offset int  option, in order to be abl…

### LINKED ISSUE #12: Port specification
The other day I was running on np04 machines and another user was running at the same time, on the same machine. 
Since it is not possible to use the same ports, launching nanorc was simply failing. 
Can we have the ports to be configurable? Even auto-configurable maybe?

### PR DESCRIPTION
**Pull request by carloschavez**
with option --port-offset <int>   one can now run two or more nanorc instances on the same node

### CODE REVIEW

**Comment by bieryAtFnal** [comment]
I'm going to add Eric as a reviewer on this Issue.

Eric, I'd appreciate it if you would test the behavioral aspects of the changes in this PR. That is, can we now run multiple nanorc instances on the same computer, if we specify different port offsets to them.  And, do the changes not affect any existing functionality.

Carlos mentioned that Alessandro would like to review this Issue from the perspective of looking at the code changes themselves.  That's great, and we should wait for that to happen before closing/merging this PR.

**Comment by eflumerf** [comment]
Existing functionality appears to be unaffected, and running with and without the option separately seem to work. Issues were encountered when trying to run both simultaneously, as the ZMQ ports were not offset.

**Comment by bieryAtFnal** [comment]
OK, good point.(ZMQ sockets were not offset).  I presume that you tried running two demo DAQ systems and noticed the problem.
There may still be value in this PR, however.  At the ICEBERG teststand, we need to run one nanorc instance for the global timing partition and one for the DAQ partition.  The changes in this PR may allow us to do that (with no additional changes).  I will run a test.

**Comment by eflumerf** [comment]
I added da9a5a3 which does the port offsetting for ZMQ, and tested with `nanorc --port-offset 1000 mdapp_2x2_swtpg_tpsw boot init conf start 101 wait 60 stop scrap terminate & nanorc mdapp_2x2_swtpg_tpsw boot init conf start 102 wait 60 stop scrap terminate`

**Comment by bieryAtFnal** [comment]
and, I fixed a typo (port-offset vs. port_offset) to get _nanotimingrc_ to work...

**Comment by bieryAtFnal** [comment]
Carlos, Pierre, Eric, and I had a quick and productive discussion about this topic just now.  Some notes:

1. We learned from Pierre that he and Alessandro had talked about this Issue, and their conclusion was that the port offset should be based on the partition number (e.g. port_offset = partno*1000) rather than a --port-offset command-line parameter.
2. I mentioned the use case of communicating with the the global timing partition, in which case the ports to/from that partition should *not* be offset
3. We talked about a couple of different ways to indicate that ports associated with the global timing partition should not be offset (changes in nanorc to notice the global timing partition and react accordingly [this seemed like a hack, so we didn't like it much], a new field in each connection specification that would tell nanorc whether it should apply a port offset of not).  Carlos and Eric will look into other possible ways that this could work.
4. Our sense was that there definitely is benefit in getting this functionality into nanorc for 2.11.0, and that the changes should not be too difficult, once we decide how to implement them.  

And, as a summary, I'll say that I think that the two primary use cases are 1) multiple developers want to run a DAQ system on the same computer and have their nanorc and DAQ connections not collide and 2) some fraction of the developers in use case 1 want to also include the global timing partition.

**Comment by carloschavez** [comment]
I added more changes [0b946ce](https://github.com/DUNE-DAQ/nanorc/pull/68/commits/0b946ce941c53026ca437cccd6a54510b2c6ae2a)  which implements changes to handle  partition_number and partition_label and now   port_offset = 1000 * partition_number  as required.

**Comment by carloschavez** [comment]
Hi @bieryAtFnal  @eflumerf  ,
This PR should be compiled/tested together with https://github.com/DUNE-DAQ/networkmanager/pull/16  and https://github.com/DUNE-DAQ/daqconf/pull/31

With these PRs/changes we are getting closer to addressing the two main use cases mentioned above , being 1) multiple developers want to run a DAQ system on the same computer and have their nanorc and DAQ connections not collide and 2) some fraction of the developers in use case 1 want to also include the global timing partition.

File: src/nanorc/__main_np04__.py [hunk 1]
```diff
@@ -39,11 +39,12 @@
 @click.option('--cfg-dumpdir', type=click.Path(), default="./", help='Path where the config gets copied on start')
 @click.option('--dotnanorc', type=click.Path(), default="~/.nanorc.json", help='A JSON file which has auth/socket for the DB services')
 @click.option('--kerberos/--no-kerberos', default=False, help='Whether you want to use kerberos for communicating between processes')
+@click.option('--port-offset', type=int, default=0, help='Application port offsetting for running more than one nanorc instance on the same node')
 @click.argument('cfg_dir', type=click.Path(exists=True))
 @click.argument('user', type=str)
 @click.pass_obj
 @click.pass_context
-def np04cli(ctx, obj, traceback, loglevel, elisa_conf, log_path, timeout, cfg_dumpdir, dotnanorc, kerberos, cfg_dir, user):
+def np04cli(ctx, obj, traceback, loglevel, elisa_conf, log_path, timeout, cfg_dumpdir, dotnanorc, kerberos, port_offset, cfg_dir, user):
```
Comment (eflumerf) [inline src/nanorc/__main_np04__.py]: This needs updated.

**Comment by bieryAtFnal** [comment]
I need to talk with ICEBERG folks to get the TLU/HSI electronics reset, and then I will be able to test these changes at ICEBERG...

**Comment by bieryAtFnal** [comment]
@carloschavez , it seems like the --partition-number option is only available for _nanorc_, not _nanotimingrc_ nor _nano04rc_.  Could it be added to all of them?
Thanks,
Kurt

**Comment by plasorak** [comment]
Hi @bieryAtFnal, I agree that --partition-number should be in `nano04rc`, however for `nanotimingrc`, since this is now a partition chooser, and the timing RC is effectively the global partition, I don't think it can be there. If really needed we can add a port offset, but my preferred solution here would be to stick with using `nanotimingrc` without any offset, and use the first partition of the `nano(04)rc`, which has a port offset.

**Comment by bieryAtFnal** [comment]
OK, that's fine.  
I'd appreciate seeing a few official words (documentation) on what we should do when running the global timing nanorc and one or more DAQ nanorc instances on the same computer, though.  When I tried partition number zero for the DAQ partition at ICEBERG when I had a global timing partition already running, things failed because of port(s) already in use, and it would have been nice to know that I shouldn't have tried to use DAQ partition number zero.

**Comment by plasorak** [comment]
Indeed, I think the solution we agreed on in the call was to have a partition number for the timing rc as well. So we'll just do that and it should fix your problem

**Comment by bieryAtFnal** [comment]
I ran a successful test at ICEBERG in which I had both the global partition and DAQ partition `nanorc` instances on the same computer.  As we discussed, I used `--partition-number 1` for the DAQ partition.
I can re-run the test, if/when there are additional code changes.

**Comment by plasorak** [comment]
@bieryAtFnal, if you have still some time, I'd appreciate another test on that branch. Now you shouldn't have to specify anything to run timing and standard nanorc on the same host. You should also be able to run several timing, np04 and standard nanorcs with `--partition-number` on the same host.

**Comment by bieryAtFnal** [comment]
Things generally look quite nice.  As you described, I can run both `nanorc` and `nanotimingrc` with the default `partition-number` values and things just work.
A couple of minor things that I noticed:  `nanotimingrc` doesn't seem to like config directories with the trailing slash, and `nanotimingrc` doesn't seem to like `--partition-number` values like 12 or 13.
In both case, the answer might be "just don't do that", which is fine.

For reference, 

- `nanorc dir1` works for me
- `nanorc dir1/` also works for me
- `nanotimingrc dir2` works for me
- `nanotimingrc dir2/` doesn't work for me, complains about `RuntimeError: '' does not exist or is not a directory`

**Comment by plasorak** [comment]
OK, for the partition number, I'll add a validation between 0-10. Definitely I'll fix that trailing slash, that's very bad.

**Pull request merged**